soc: arm: nxp_imx: rt: enable SOC fixed MPU regions, add SDRAM0 region

danieldegrasse opened the pull request:
Enable SOC fixed MPU regions by default for the RT10xx/RT11xx SOC lines.

Additionally, add code to handle defining the SDRAM0 region as
device type (non cacheable, non shareable), when no SDRAM0 node is
present. This will resolve an issue present on the RT11xx series where
the core may execute speculative prefetches to the SDRAM region when
no SDRAM is present on the board, resulting in the system faulting.

This PR also adds documentation to the RT series boards to highlight the external memories present. This is intended to
improve the customer experience when booting a custom board with Zephyr

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: arm: nxp_imx: rt: enable SOC fixed MPU regions, add SDRAM0 region

Enable SOC fixed MPU regions by default for the RT10xx/RT11xx SOC lines.

Additionally, add code to handle defining the SDRAM0 region as
device type (non cacheable, non shareable). This behavior can
be disabled with CONFIG_NXP_IMX_EXTERNAL_SDRAM=y. Set this Kconfig
for all boards in tree using SDRAM.

This will resolve an issue present on the RT11xx series where
the core may execute speculative prefetches to the SDRAM region when
no SDRAM is present on the board, resulting in the system faulting.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: arm: nxp_imx: rt: Remove CONFIG_OCRAM_NOCACHE setting

Remove CONFIG_OCRAM_NOCACHE setting, as this is now possible to achieve
using devicetree linker regions, and there is no point in having a
specific Kconfig for one memory region on the RT series like this.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

decsny commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -9,9 +9,9 @@ zephyr_sources_ifdef(CONFIG_SOC_SERIES_IMX_RT10XX soc_rt10xx.c)
 
-# Enable custom OCRAM noncacheable region
+# Add custom mpu regions
+zephyr_sources(mpu_regions.c)
```
maybe something should go in the release notes about this change?

danieldegrasse replied:
Sure- I'll highlight this in the next update of this PR

danieldegrasse committed:
soc: arm: nxp_imx: don't default CONFIG_DEVICE_CONFIGURATION_DATA=y

Don't default CONFIG_DEVICE_CONFIGURATION_DATA to enabled for iMX RT
SOCs, as this configuration block is only used when the board needs
peripherals like the external memory controller setup from reset by the
bootrom.

Enable this feature on all in tree boards that will require it,
and document the change to the default value in release notes.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: arm: nxp: add documentation to RT series about external memories

Add documentation to NXP i.MX RT series SOCs about external memories
present on the board. This is intended to help highlight how the SOC
enables external memory using configuration blocks exposed to the
bootrom.

Addtionally, add a check to these boards to warn a user if they are
attempting to utilize SDRAM as their RAM bank without a DCD present.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented on soc/arm/nxp_imx/rt/mpu_regions.c:
```diff
@@ -1,37 +1,37 @@
+	 * no SDRAM is present.
+	 */
+	MPU_REGION_ENTRY("SDRAM0",
+			SDRAM_BASE_ADDR, REGION_IO_ATTR(REGION_512M)),
```
Yes, this is intended to override the default attributes assigned to the 512M region starting at `0x8000_0000` in the ARMv7-M system memory map: 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/b1a59746-0006-4cd1-8c90-bd780b3c8ba4)

Indentation has been fixed in latest push

danieldegrasse commented:
> @pabigot, since you did look at this PR, can you provide an approval?

Think you might have meant @pdgendt :)
